sql-plugin/src/main/scala/com/nvidia/spark/rapids/RapidsGdsStore.scala

```diff
@@ -230,26 +236,62 @@ class RapidsGdsStore(
     class RapidsGdsBatchedBuffer(
         id: RapidsBufferId,
         path: File,
-        fileOffset: Long,
-        size: Long,
+        val fileOffset: Long,
+        override val size: Long,
         meta: TableMeta,
         spillPriority: Long,
         spillCallback: RapidsBuffer.SpillCallback,
         var isPending: Boolean = true)
         extends RapidsGdsBuffer(id, size, meta, spillPriority, spillCallback) {
 
+      private def batchUnspill(buffer: DeviceMemoryBuffer): Unit = {
+        withResource(buffer) { _ =>
+          spilledBuffers.get(path).foreach { id =>
+            withResource(catalog.acquireBuffer(id)) {
+              case gdsBuffer: RapidsGdsBatchedBuffer =>
+                closeOnExcept(DeviceMemoryBuffer.allocate(gdsBuffer.size)) { d =>
+                  logDebug(s"Unspilling $gdsBuffer $id to $DEVICE")
+                  d.copyFromMemoryBuffer(
+                    0, buffer, gdsBuffer.fileOffset, gdsBuffer.size, Cuda.DEFAULT_STREAM)
+                  try {
+                    deviceStorage.copyBuffer(gdsBuffer, d, Cuda.DEFAULT_STREAM)
+                  } catch {
+                    case _: DuplicateBufferException =>
+                      logDebug(s"Lost device buffer registration race for buffer $id, ignored")
+                  }
+                }
+              case _ =>
+                logDebug(s"Ignoring buffer $id")
+            }
+          }
+        }
+      }
+
       override def materializeMemoryBuffer: MemoryBuffer = this.synchronized {
-        closeOnExcept(DeviceMemoryBuffer.allocate(size)) { buffer =>
-          if (isPending) {
+        if (isPending) {
```

Comment:
Could there be some comments in `materializeMemoryBuffer` for each of the three cases?

Reply:
Done.